pkg/noderesourcetopology/filter.go

```diff
@@ -0,0 +1,254 @@
+
+// TopologyMatch plugin which run simplified version of TopologyManager's admit handler
+type TopologyMatch struct {
+	policyHandlers PolicyHandlerMap
+	lister         listerv1alpha1.NodeResourceTopologyLister
+	namespaces     []string
+}
+
+type NUMANode struct {
+	NUMAID    int
+	Resources v1.ResourceList
+}
+
+type NUMANodeList []NUMANode
+
+// Name returns name of the plugin. It is used in logs, etc.
+func (tm *TopologyMatch) Name() string {
+	return Name
+}
+
+func extractResources(zone topologyv1alpha1.Zone) v1.ResourceList {
+	res := make(v1.ResourceList)
+	for _, resInfo := range zone.Resources {
+		quantity, err := resource.ParseQuantity(resInfo.Allocatable.String())
+		if err != nil {
+			klog.Errorf("Failed to parse %s", resInfo.Allocatable.String())
+			continue
+		}
+		res[v1.ResourceName(resInfo.Name)] = quantity
+	}
+	return res
+}
+
+func SingleNUMAContainerLevelHandler(pod *v1.Pod, zones topologyv1alpha1.ZoneList) *framework.Status {
+	klog.V(5).Infof("Single NUMA node handler")
+
+	// prepare NUMANodes list from zoneMap
+	nodes := createNUMANodeList(zones)
+	qos := v1qos.GetPodQOS(pod)
+	for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
+		resBitmask := checkResourcesForNUMANodes(nodes, container.Resources.Requests, qos)
+		if resBitmask.IsEmpty() {
+			// definitely we can't align container, so we can't align a pod
+			return framework.NewStatus(framework.Unschedulable, fmt.Sprintf("Cannot align container: %s", container.Name))
+		}
+	}
+	return nil
+}
+
+func checkResourcesForNUMANodes(nodes NUMANodeList, resources v1.ResourceList, qos v1.PodQOSClass) bm.BitMask {
+	bitmask := bm.NewEmptyBitMask()
+	bitmask.Fill()
+
+	zeroQuantity := resource.MustParse("0")
+	for resource, quantity := range resources {
+		resourceBitmask := bm.NewEmptyBitMask()
+		for _, numaNode := range nodes {
+			numaQuantity, ok := numaNode.Resources[resource]
+			// if can't find requested resource on the node - skip (don't set it as available NUMA node)
+			// if unfound resource has 0 quantity probably this numa node can be considered
+			if !ok && quantity.Cmp(zeroQuantity) != 0 {
+				continue
+			}
+			// Check for the following:
+			// 1. set numa node as possible node if resource is memory or Hugepages (until memory manager will not be merged and
+			// memory will not be provided in CRD
+			// 2. set numa node as possible node if resource is cpu and it's not guaranteed QoS, since cpu will flow
+			// 3. set numa node as possible node if zero quantity for non existing resource was requested (TODO check topology manaager behaviour)
+			// 4. otherwise check amount of resources
+			if resource == v1.ResourceMemory ||
+				strings.HasPrefix(string(resource), string(v1.ResourceHugePagesPrefix)) ||
+				resource == v1.ResourceCPU && qos != v1.PodQOSGuaranteed ||
+				quantity.Cmp(zeroQuantity) == 0 ||
+				numaQuantity.Cmp(quantity) >= 0 {
+				resourceBitmask.Add(numaNode.NUMAID)
```

Comment:
IIUC, you can directly use the top-level `bitmask` to manipulate, i.e., eliminate `resourceBitmask`.

Reply:
It could be done in this way too, e.g. by bitmask.Remove

Reply:
The previous version is more readable, sorry for the confusion - I misread `Add` and `And`, so thought `resourceBitmask` was unneeded. Please revert to the previous implementation, and:

- add some comments around `resourceBitmask` like: "for each requested resource, calculate which NUMA slots are good fits, and then AND with the aggregated bitmask".
- we can change signature of `checkResourcesForNUMANodes` to return a boolean value, so that to return early when bitmask is empty - after every iteration it ANDs with `resourceBitmask`. WDYT?

Reply:
Bump. Please add some comments for `resourceBitmask`, and rename `checkResourcesForNUMANodes` to `resMatchNUMANodes(...) bool`.

Reply:
Bump again.